contracts/Oracle.sol

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+contract BalancerPriceGuard {
+    struct PriceSnapshot {
+	    uint256 cumulativePrice; // Time-weighted cumulative price
+    	    uint256 lastUpdated;     // Timestamp of the last update
+            uint256 averagePrice;    // Most recent calculated average price
+    }
+
+    PriceSnapshot public snapshotHistory;
+    uint256 public constant maxSlippageUpdate = 10; // Maximum allowed update slippage in percentage
+
+    constructor(uint256 _maxSlippageUpdate) {
+        require(_maxSlippage <= 10000, "Slippage must be <= 100%");
+        maxSlippageUpdate = _maxSlippageUpdate; // Default slippage in basis points (e.g., 500 for 5%)
+	updatePrice(pool, tokenIn, tokenOut); // trusted update 
+    }
+
```

Comment:
Is all this code borrowed from somewhere? If so, please highlight the source

Reply:
see above. just ignore Oracle.sol in this PR.